### ISSUE TITLE
IOP export table declaration causes relocation truncated warning when mgpopt is enabled

### ISSUE THREAD

**Issue by sp193** [ps2dev/ps2sdk#73]
As of commit 587cad4, these macros have been added to defs.h, to allow the programmer to manipulate the $gp register when required:

- SetGP
-GetGP
-ReplaceGP
-SetModuleGP

USBD will now use $gp, and also supports restoring the driver & caller's $gp. This was a documented function of USBD, which our homebrew module was lacking.

However, the G parameter is currently set to 19. I found that I couldn't increase it, as with the SMAP module.
Going beyond 19 would result in a linker error like this:
``` iop-gcc -D_IOP -fno-builtin -Os -G0 -Wall  -Isrc/ -Isrc/include -Iinclude/ -I/home/nozomi/PS2/ps2sdk/iop/kernel/include -I/home/nozomi/PS2/ps2sdk/common/include -mgpopt -G8192 -o irx/usbd.irx obj/hcd.o obj/hub.o obj/interface.o obj/mem.o obj/usbd.o obj/usbio.o obj/driver.o obj/imports.o obj/exports.o -nostdlib -s
obj/usbd.o(.text+0x700): In function `_start':
src/usbd.c: relocation truncated to fit: R_MIPS_GPREL16 _exp_usbd
collect2: ld returned 1 exit status
```

I think it might have something to do with the [definition of the export table](https://github.com/ps2dev/ps2sdk/blob/master/iop/kernel/include/irx.h#L85), which has the fptrs field defined as an zero-length array at +0x14 (20th byte mark).
Removing the declaration of the _exp_usbd structure and its corresponding call to RegisterLibraryEntries will inhibit the warning, even if I increased the G-parameter to as high as 8192.

And so I think it might have got to do with the undefined size of fptrs, but I am not well-versed with the GCC/LD internals.

For now, I have a patch that replaces the use of struct irx_export_table with iop_libhead_t, which represents the export table after linking: https://github.com/sp193/ps2sdk/tree/libhead
Since it does not define the export table body (fptrs), it results in no linker error.
Other than solving this problem, it also represents the import and export tables after linking (rather than their forms before linking). So these new structures can be used together with the ones in irx.h for complete representation of the import and export tables.

The only bad thing about doing this, is that there will be warnings in older projects, but I guess that is a small price to pay.

Does anybody have any comments or a better way to solve this?

**Comment by rickgaiser** [comment]
Do we really need to use $gp register? I've had a hard to debug problem that was coused by the use of the $gp register in smap.irx. Removing the use of this register solved my problems. Also when upgrading to newer gcc version (if this is ever going to happen) will have conflicts with this method of using the $gp register, making the upgrade even harder than it already is.

Another thing is that it complicates development of .irx modules, and increases the chance of bugs when the $gp register is not restored correctly.

Is there a really good reason to go through all this trouble? What is the benefit of using the $gp register?

**Comment by sp193** [comment]
You get one more register to use, since go was made for accessing global and static data.

But for USBD, it is more because it was a documented feature, whereby the programmer can expect $gp to remain correct in callbacks.

**Comment by sp193** [comment]
These are SDK modules, so perhaps Sony wanted to cater for all uses.

The $gp register is only preserved between context switches and certain callback mechanisms. USBD would be one module which supports restoring $gp for any driver or callback.
However, each module may have a different value for $gp. So if it is not updated between inter-module calls, it will become wrong. Your driver (and all others too) module will have a different $gp value from USBD's callback thread.

**Comment by rickgaiser** [comment]
So there's actually 2 situations to consider:
1 - Supporting modules that use $gp, by restoring the client value of $gp before calling back the client. This makes sense for compatibility with proprietary modules that use $gp.
2 - Using the $gp register in ps2sdk modules. Meaning every exported function call must save/restore the $gp register manually before $gp can be used inside that module.

Isn't it possible to do 1, without doing 2?

**Comment by sp193** [comment]
Yes. And yes, it is possible if you choose to not use $gp.

I plan to only make the kernel modules function exactly as how Sony designed them to, so that they will function exactly as the original did.
USBD does have a large global structure. So by using $gp, I think they also hoped to reduce the number of load immediate operations that are required for accessing the globals, from across all functions.

But the purpose of my question is about how we can make the mgopt option work seriously with our declaration of the export table. Or if the export table structure's design is just a problematic design, which should be replaced.
Right now, everything is in place and the use of the $gp register within USBD should work. It's just that I have been unable to set the G-parameter higher since 2013. Hence the use of the $gp register is possibly limited within SMAP and USBD, to really small pieces of global data.

The Sony version of the export table is like the new libhead structure, which has no body. They use a program to generate the export table, so they likely don't use a structure that contains a zero-length element like ours.
The libhead structure also represents the tables after linking, while our structure represents the structures before linking. So there is also some slight difference in purpose, which might be desirable.
Normally, I would just merge in such a change, but I haven't decided if it is worth all the compiler warnings in older software (mismatched arguments).

**Comment by sp193** [comment]
I seem to have blundered somewhere again. :(
Removing those fields (through new structure or not) doesn't help! So I have no idea why I get such linker errors.

As for the fptrs field and friends: I seem to have been mistaken again. Not only does omitting it not help, you're probably right about it not being used (the macros for generating import & export tables work differently)... so everything here has been a waste of time.

As for the complexity: if used right, it would reduce the number of immediate loads **very** slightly. I have observed that programs compiled with mgpopt are usually slightly shorter too.

It is especially useful if the function will access a global structure across **multiple** function calls.
For example, take [smap_init of smap.c](https://github.com/ps2dev/ps2sdk/blob/master/iop/network/smap/src/smap.c#L837). It has a series of strcmp operations, with the string literal being a static constant.

Without mgpopt, the full address of the globals and constants must be loaded (with lui and addiu/ori) for each evaluation.
```    1604:	8e050000 	lw	a1,0(s0)
    1608:	3c040000 	lui	a0,0x0
    160c:	0c000000 	jal	0 <strcmp>
    1610:	2484041c 	addiu	a0,a0,1052
    1614:	14400128 	bnez	v0,1ab8 <smap_init+0x524>
    1618:	24020001 	li	v0,1
    161c:	3c010000 	lui	at,0x0
    1620:	ac220008 	sw	v0,8(at)
```
With mgpopt:
```    157c:	8e050000 	lw	a1,0(s0)
    1580:	2784c0d0 	addiu	a0,gp,-16176
    1584:	0c000000 	jal	0 <strcmp>
    1588:	00000000 	nop
    158c:	14400127 	bnez	v0,1a2c <smap_init+0x520>
    1590:	24020001 	li	v0,1
    1594:	af82c018 	sw	v0,-16360(gp)
```

Of course, if you replaced $gp in a function that does not refer to many globals, then the trade-off becomes minimal (might even add onto the complexity).

Thanks for discussing this matter with me, even thought it didn't exactly lead to anything we can use directly yet.

**Comment by rickgaiser** [comment]
I've read that using $gp also requires a special linker script. The global data needs to be inside a 64K area, and _gp pointing to exactly the middle (start + 0x8000)? Perhaps this is what's causing the linker error.

**Comment by sp193** [comment]
Thanks, I just remembered why. I have seen this problem and fixed it before, within FMCBInstaller.
Although the reason is different.

In a lot of my tools, I always use the $gp register. However, I got this same error as FMCBInstaller needed to access the errno global variable.
The libraries are compiled with mno-gpopt, hence errno is placed in .data instead of .sdata. Hence exceeding the permissible limitations of the relocation.

Here, it is because we declare the export table header as being only 20 bytes. That makes GCC think we can put it in .sdata... but the real export table is too big, and is put into .data.
So I have added attributes that tell GCC that the export table is in .data. I have also increased the G-option to 16384.
